### PR TITLE
Statically type various parts of bokeh

### PR DESCRIPTION
**Pull request by mattpap**
I started with `bokeh.application.handlers` and then gradually typed everything that was imported there and so on. As with other related PRs, this is a rough approximation, which will be gradually refined as type coverage increases.

### CODE REVIEW

File: bokeh/_testing/util/filesystem.py [hunk 1]
```diff
@@ -143,20 +154,12 @@ async def with_temporary_file_async(func, dir=None, suffix=''):
...
-
-    def __enter__(self):
-        os.chdir(self._new)
-        return self._new
+@contextmanager
+def WorkingDir(new: PathLike) -> Iterator[PathLike]:
+    old = os.getcwd()
+    os.chdir(new)
+    yield new
+    os.chdir(old)
```
Comment (bryevdv) [inline bokeh/_testing/util/filesystem.py:162]: @mattpap I thought we recently agreed on a different style here https://github.com/bokeh/bokeh/pull/11219#discussion_r623270110

```
def _get_save_args(
    state: State, 
    filename: Optional[str], 
    resources: Optional[Resources],
    title: Optional[str]
) -> Tuple[str, Resources, str]:
```

File: bokeh/_testing/util/filesystem.py [hunk 2]
```diff
@@ -55,35 +66,35 @@ def __exit__(self, type, value, traceback):
         except Exception as e:
             # prefer original exception to rmtree exception
             if value is None:
-                print("Exception cleaning up TmpDir %s: %s" % (self._dir, str(e)), file=sys.stderr)
+                print(f"Exception cleaning up TmpDir {self._dir}: {str(e)}", file=sys.stderr)
```
Comment (bryevdv) [inline bokeh/_testing/util/filesystem.py:69]: `str` is redundant here
```python
In [3]: try:
   ...:     raise RuntimeError("foo")
   ...: except RuntimeError as e:
   ...:     print(f"{e}")
   ...:
foo
```

File: bokeh/application/handlers/code_runner.py [hunk 3]
```diff
@@ -87,11 +102,12 @@ def __init__(self, source, path, argv, package=None):
...
 
         try:
-            nodes = ast.parse(source, path)
+            nodes = ast.parse(source, os.fspath(path))
             self._code = compile(nodes, filename=path, mode='exec', dont_inherit=True)
         except SyntaxError as e:
             self._code = None
-            self._permanent_error = ("Invalid syntax in \"%s\" on line %d:\n%s" % (os.path.basename(e.filename), e.lineno, e.text))
+            filename = os.path.basename(e.filename) if e.filename is not None else "???"
+            self._permanent_error = f"Invalid syntax in \"{filename}\" on line {e.lineno or '???'}:\n{e.text or '???'}"
```
Comment (bryevdv) [inline bokeh/application/handlers/code_runner.py:110]:
```suggestion
            self._permanent_error = f"Invalid syntax in {filename!r} on line {e.lineno or '???'}:\n{e.text or '???'}"
```

File: bokeh/application/handlers/handler.py [hunk 4]
```diff
@@ -217,14 +237,14 @@ def url_path(self):
...
 
     '''
     handler._failed = True
     handler._error_detail = traceback.format_exc()
 
-    _exc_type, _exc_value, exc_traceback = sys.exc_info()
+    _, _, exc_traceback = sys.exc_info()
     filename, line_number, func, txt = traceback.extract_tb(exc_traceback)[-1]
 
     handler._error = "%s\nFile \"%s\", line %d, in %s:\n%s" % (str(e), os.path.basename(filename), line_number, func, txt)
```
Comment (bryevdv) [inline bokeh/application/handlers/handler.py:250]: as long as you are fixing up f-srings in these files here is another

File: bokeh/client/states.py [hunk 5]
```diff
@@ -50,39 +58,45 @@ class ErrorReason(Enum):
     HTTP_ERROR      = auto()
     NETWORK_ERROR   = auto()
 
-class NOT_YET_CONNECTED:
+class State(metaclass=ABCMeta):
```
Comment (bryevdv) [inline bokeh/client/states.py:61]: Perhaps this would be better now as a `Protocol`?

File: bokeh/model.py [hunk 6]
```diff
@@ -271,10 +298,10 @@ def __str__(self) -> str:
     __repr__ = __str__
 
     @property
-    def id(self) -> str:
+    def id(self) -> ID:
         return self._id
 
-    name: tp.Union[None, str] = Nullable(String, help="""
+    name: str | None = p.Nullable(p.String, help="""
```
Comment (bryevdv) [inline bokeh/model.py:304]: What's the rationale for sprinkling `p.` everywhere?

File: bokeh/client/states.py (hunk 5 above)
Comment (mattpap) [reply bokeh/client/states.py:62]: Not really. Actually both approaches don't give any static feedback whether sub-classes are implemented correctly or not (no wonder as there's no `@abstract` decorator; ours does nothing). At least with `ABC` we will get a meaningful exception at runtime.

File: bokeh/_testing/util/filesystem.py (hunk 1 above)
Comment (mattpap) [reply bokeh/_testing/util/filesystem.py:162]: I applied this to some extreme cases, but otherwise it's a lot of rewriting at this point. Let's do that after the initial typing pass is done (which I think can be achieved before 2.4).

**Comment by mattpap** [comment]
I will stop here and continue with another typing round in another PR.

File: bokeh/model.py (hunk 6 above)
Comment (bryevdv) [reply bokeh/model.py:304]: AFAIK the only thing that makes use of the "contribute multiple descriptors" functionality is the automatic `_units` properties. If those are cleaned up there's probably no reason the properties themselves couldn't just become the descriptors directly, rather than descriptor factories, and that would be a big simplification.

File: tests/unit/bokeh/test___init__.py [hunk 7]
```diff
@@ -78,7 +80,7 @@ def test___version___type() -> None:
 def test___version___defined() -> None:
     assert b.__version__ != 'unknown'
 
-def test_license(capsys) -> None:
+def test_license(capsys: Capture) -> None:
```
Comment (bryevdv) [inline tests/unit/bokeh/test___init__.py:83]: What benefit is provided by typing pytest fixtures?

Comment (mattpap) [reply tests/unit/bokeh/test___init__.py:83]: The benefit can be seen two lines later:
```py
out, err = capsys.readouterr()
assert out == _LICENSE
```
Now you know what `capsys` is and what's its API, and you don't have to run this code or search for "pytest capsys" to know. In fact, I hate this implicit fixture API, as it's often unclear what the thing is and where it came from (is it pytest's or thirpardy). I'm pretty sure some IDEs are clever and can special-case such features, but I prefer this to be self-evident.

File: bokeh/util/serialization.py [hunk 8]
```diff
@@ -159,9 +196,9 @@ def convert_timedelta_type(obj):
     if isinstance(obj, dt.timedelta):
         return obj.total_seconds() * 1000.
     elif isinstance(obj, np.timedelta64):
-        return (obj / NP_MS_DELTA)
+        return float(obj / NP_MS_DELTA)
 
-def convert_datetime_type(obj):
+def convert_datetime_type(obj: pd.NaT | pd.Period | pd.Timestamp | pd.Timedelta | dt.datetime | dt.date | dt.time | np.datetime64) -> float:
```
Comment (bryevdv) [inline bokeh/util/serialization.py:201]: type alias seems appropriate here

File: bokeh/server/callbacks.py [hunk 9]
```diff
@@ -158,44 +175,44 @@ def __init__(self, document, callback, timeout, id=None):
...
+    def _copy_with_changed_callback(self, new_callback: Callback) -> TimeoutCallback:
         ''' Dev API used to wrap the callback with decorators. '''
         return TimeoutCallback(self._document, new_callback, self._timeout, self._id)
 
 #-----------------------------------------------------------------------------
 # Private API
 #-----------------------------------------------------------------------------
 
-class _DocumentCallbackGroup:
+class DocumentCallbackGroup:
```
Comment (bryevdv) [inline bokeh/server/callbacks.py:192]: Needs to be moved above Private API comment line if so

File: bokeh/resources.py [hunk 10]
```diff
@@ -594,23 +624,23 @@ class Resources(JSResources, CSSResources):
...
+        return f"{css}\n{js}"
 
 
 # -----------------------------------------------------------------------------
 # Private API
 # -----------------------------------------------------------------------------
 
-
-class _SessionCoordinates:
+class SessionCoordinates:
```
Comment (bryevdv) [inline bokeh/resources.py:636]: Move out of Private API section

File: bokeh/resources.py [hunk 11]
```diff
@@ -671,50 +702,47 @@ def _get_cdn_urls(version=None, minified=True, legacy=False):
...
-                "type": "warn",
-                "text": (
-                    "Requesting CDN BokehJS version '%s' from Bokeh development version '%s'. "
-                    "This configuration is unsupported and may not work!" % (version, __version__)
-                ),
-            }
-        )
-
-    if is_full_release(version):
+        result.messages.append(Message(
```
Comment (bryevdv) [inline bokeh/resources.py:714]: "Message" is awfully generic. Can we make this type more specific? AFAIK this is only / mostly used to publish messages to notebook output cells but it might be OK to denote it as a "JS runtime message" in some way

File: bokeh/resources.py [hunk 12]
```diff
@@ -671,50 +702,47 @@ def _get_cdn_urls(version=None, minified=True, legacy=False):
...
+                "This configuration is unsupported and may not work!"
+            ),
+        ))
+
+    if is_full_release(version): # TODO: TypeGuard
+        assert version is not None
         sri_hashes = get_sri_hashes_for_version(version)
-        result['hashes'] = lambda components, kind: {mk_url(component, kind): sri_hashes[mk_filename(component, kind)] for component in components}
+        result.hashes = lambda components, kind: \
+            { mk_url(component, kind): sri_hashes[mk_filename(component, kind)] for component in components }
```
Comment (bryevdv) [inline bokeh/resources.py:726]: no need for explicit line continuation
```suggestion
        result.hashes = lambda components, kind: {
            mk_url(component, kind): sri_hashes[mk_filename(component, kind)] for component in components
        }
```

File: bokeh/protocol/messages/pull_doc_req.py [hunk 13]
```diff
@@ -36,26 +39,26 @@
...
+    def create(cls, **metadata: Any) -> pull_doc_req:
         ''' Create an ``PULL-DOC-REQ`` message
 
         Any keyword arguments will be put into the message ``metadata``
         fragment as-is.
 
         '''
         header = cls.create_header()
-        return cls(header, metadata, {})
+        return cls(header, metadata, Empty())
```
Comment (bryevdv) [inline bokeh/protocol/messages/pull_doc_req.py:61]: `Empty` is too general. Empty dict, tuple, string? If we need "empty" things in other contexts we will end up in the situation where there is one general name (that means something specific) and later-on more specific names, which is always an annoying inconsistency. This type should be named something more specific from the start.

File: bokeh/protocol/__init__.py [hunk 14]
```diff
@@ -102,11 +123,9 @@ def assemble(self, header_json, metadata_json, content_json):
         '''
         header = json_decode(header_json)
         if 'msgtype' not in header:
-            log.error("Bad header with no msgtype was: %r", header)
+            log.error(f"Bad header with no msgtype was: {header!r}")
             raise ProtocolError("No 'msgtype' in header")
-        return self._messages[header['msgtype']].assemble(
-            header_json, metadata_json, content_json
-        )
+        return self._messages[header['msgtype']].assemble(header_json, metadata_json, content_json)
```
Comment (bryevdv) [inline bokeh/protocol/__init__.py:128]:
```suggestion
        return self._messages[header["msgtype"]].assemble(header_json, metadata_json, content_json)
```

**Review by bryevdv — CHANGES_REQUESTED** [review]

File: tests/unit/bokeh/core/property/test_enum.py [hunk 15]
```diff
@@ -41,10 +41,10 @@ def test_init(self) -> None:
         with pytest.raises(TypeError):
             bcpe.Enum()
 
-        with pytest.raises(TypeError):
+        with pytest.raises(ValueError):
```
[the remark this answers is not in the document]
Comment (mattpap) [reply tests/unit/bokeh/core/property/test_enum.py:44]: It always was `ValueError`, but the underlying code was wrong, resulting in an unrelated `TypeError`:
```
TypeError: not all arguments converted during string formatting
```
I accidentally fixed this when switching to template literals.

File: bokeh/server/callbacks.py (hunk 9 above)
Comment (mattpap) [reply bokeh/server/callbacks.py:192]: I changed this, because this is in fact not a private (to a module) API, but used elsewhere, whereas `_` is a clear indicator of a private API. Though I'm not sure if it's a dev API as well, perhaps an internal API. If we treat `_` as an internal API indicator instead, then I can revert this. Though various linters would warn about this.

File: tests/unit/bokeh/embed/test_util__embed.py [hunk 16]
```diff
@@ -513,20 +513,7 @@ def test_passing_model(self) -> None:
...
-        p1 = SomeModel()
-        d = Document()
-        d.add_root(p1)
-        docs_json, render_items = beu.standalone_docs_json_and_render_items([d])
-        doc = list(docs_json.values())[0]
-        assert doc['title'] == "Bokeh Application"
-        assert doc['version'] == __version__
-        assert len(doc['roots']['root_ids']) == 1
-        assert len(doc['roots']['references']) == 1
-        assert doc['roots']['references'] == [{'attributes': {}, 'id': str(p1.id), 'type': 'test_util__embed.SomeModel'}]
```
[the remark this answers is not in the document]
Comment (mattpap) [reply tests/unit/bokeh/embed/test_util__embed.py]: I thought this was a duplicated test, but actually it's not.

File: bokeh/layouts.py [hunk 17]
```diff
@@ -75,7 +75,12 @@
 # General API
 #-----------------------------------------------------------------------------
 
-def row(*args: LayoutDOM, children: List[LayoutDOM] | None = None, sizing_mode: SizingModeType | None = None, **kwargs) -> Row:
+@overload
+def row(children: List[LayoutDOM], *, sizing_mode: SizingModeType | None = None, **kwargs: Any) -> Row: ...
+@overload
+def row(*children: LayoutDOM, sizing_mode: SizingModeType | None = None, **kwargs: Any) -> Row: ...
```
Comment (bryevdv) [inline bokeh/layouts.py:81]: Overloading is useful when return types vary according to input:

https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading

What is the purpose of using overloads here where there is no such relationship?

Comment (mattpap) [reply bokeh/layouts.py:81]: That's one class of signatures where overloads help. Another, as used here, helps us to disallow mutually exclusive combinations of arguments. In this case we want to disallow `row(Plot(), children=[Plot()])`. Previously this could have been verified only at run time (it still is verified).

File: release/__init__.py [hunk 18]
```diff
@@ -4,3 +4,4 @@
 #
 # The full license is in the file LICENSE.txt, distributed with this software.
 # -----------------------------------------------------------------------------
+from __future__ import annotations
```
Comment (bryevdv) [inline release/__init__.py:7]: Why is this necessary or useful? The computational cost of computing annotations at runtime is irrelevant for the release scripts. Please revert.

**Review by bryevdv — CHANGES_REQUESTED** [review]

**Comment by bryevdv** [comment]
@mattpap This PR has grown to be un-reviewable. Please remove changes that can easily be accomplished in separate PRs (and with discussions) e.g. the lazy annotations, in order to make this PR more manageable.

File: bokeh/layouts.py (hunk 17 above)
Comment (bryevdv) [reply bokeh/layouts.py:81]: > it still is verified

You have removed runtime checks in other instances, what is your decision criteria for leaving in vs removing?

Comment (mattpap) [reply bokeh/layouts.py:81]: I always leave them, because the goal of this work is not to change run time logic and at run time, overloads collapse into a single signature to, so you have to analyze this anyway. But at least, when using this function is used in a typed context, one will know what combinations make sense and which don't.

File: release/__init__.py (hunk 18 above)
Comment (mattpap) [reply release/__init__.py:7]: I added this for the sake of consistency across the codebase, so that we have the same semantics everywhere. Previously I was adding this on case-by-case basis, but tripping over differing semantics gets annoying pretty quickly. I didn't consider performance implications.

Comment (bryevdv) [reply release/__init__.py:7]: If this is just about forward references we could continue to use strings and make a clean switch without `__future__` when 3.10 is the min version, instead of adding so much diff noise now.

File: bokeh/layouts.py (hunk 17 above)
Comment (bryevdv) [reply bokeh/layouts.py:81]: looking back, the cases where you removed runtime checks was for cases where function signatures that were changed

**Comment by bryevdv** [comment]
@mattpap thanks for splitting this up. It this is `ready` now I will take a last look over it tonight

File: bokeh/application/handlers/code_runner.py [hunk 19]
```diff
@@ -188,23 +204,23 @@ def run(self, module, post_check=None):
...
+        #
+        # https://docs.python.org/2/library/sys.html#sys.path
+        _cwd = os.getcwd()
+        _sys_path = list(sys.path)
+        _sys_argv = list(sys.argv)
+        sys.path.insert(0, os.path.dirname(self._path))
+        sys.argv = [os.path.basename(self._path)] + self._argv
 
+        try:
+            assert self._code is not None
```
[the remark this answers is not in the document]
Comment (mattpap) [reply bokeh/application/handlers/code_runner.py:217]: This assertion is for aiding type checker only. Currently `_code` can be `None` on the type-level, but I expect this shouldn't happen at run time, unless `except Exception` obscures this. Eventually this line should be removed. However,  I did overlook that `Exception` is being caught, which would nullify this assertion at run time.

File: bokeh/models/sources.py [hunk 20]
```diff
@@ -180,36 +206,41 @@ class ColumnDataSource(ColumnarDataSource):
...
-    def column_names(self):
+    def column_names(self) -> TList[str]:
         ''' A list of the column names in this data source.
 
         '''
         return list(self.data)
 
     @staticmethod
-    def _data_from_df(df):
+    def _data_from_df(df: pd.DataFrame) -> DataDict:
```
[the remark this answers is not in the document]
Comment (bryevdv) [reply bokeh/models/sources.py:243]: oh I suppose this is the `__future__` import

[the remark this answers is not in the document]
Comment (mattpap) [reply bokeh/models/sources.py:243]: This is the magic of `from __future__ import annotations`, e.g.:
```py
In [1]: from __future__ import annotations

In [2]: pd = None

In [3]: def foo(bar: pd.DataFrame): pass

In [4]: foo("whatever")
```
Previously one could use `def foo(bar: "pd.DataFrame")` to achieve the same. In both cases, this `pd` is not going to be evaluated at run time, and only analyzed by type checkers.

File: bokeh/client/websocket.py [hunk 21]
```diff
@@ -44,17 +47,15 @@
 class WebSocketClientConnectionWrapper:
     ''' Used for compatibility across Tornado versions and to add write_lock'''
 
-    def __init__(self, socket):
-        if socket is None:
-            raise ValueError("socket must not be None")
```
[the remark this answers is not in the document]
Comment (mattpap) [reply bokeh/client/websocket.py:49]: For the most part yes, especially in public user facing APIs, but here I don't see a need for one. First, because this is internal API and won't be used interactively (e.g. in IPython), so we can rely on the type system to tell us if we are using it incorrectly. Second, if this is supposed to be fully useful, then and `isinstance()` check would be my preferred way to proceed, but given the first point, I would consider this redundant.

[the remark this answers is not in the document]
Comment (mattpap) [reply bokeh/client/websocket.py:49]: In general, we will need to establish a pattern for which APIs we rely solely on the type system and were we need both. However, there are libraries that can use static type annotations and generate run-time type checks, so we may want to consider those at some point, to reduce redundancy and a possibility of mismatch between those two types of type checking.

**Comment by mattpap** [comment]
I think I addressed most of the concerns. There are some comments that are/will be resolved in PR #11330 (e.g. `bokeh.layouts`). I didn't resolve `Message[Empty]` naming convention yet, as this may need to be redesigned further and `Empty` removed altogether.

File: bokeh/embed/bundle.py [hunk 22]
```diff
@@ -59,66 +76,72 @@
...
-class Style:
-    def __init__(self, content):
+class Style(Artifact):
+    def __init__(self, content: str):
         self.content = content
 
 
 class Bundle:
-    @classmethod
-    def of(cls, js_files, js_raw, css_files, css_raw, hashes):
```
Comment (philippjfr) [inline bokeh/embed/bundle.py:87]: @mattpap Any reason this was removed here without a note? I do find this method a bit weird (and superfluous) but since it was part of the API and used in a few places Panel used it and it's now breaking. Happy to see this removed as a proper deprecation but not as part of a PR which is meant to be purely about typing.

Comment (philippjfr) [reply bokeh/embed/bundle.py:87]: Actually don't worry about it, I'll just patch it in Panel. A heads up when you change APIs (even of the internals) would still be appreciated.

Comment (mattpap) [reply bokeh/embed/bundle.py:87]: It's such an obscure API that I didn't bother to check if panel doesn't use it (I usually do check).

Comment (mattpap) [reply bokeh/embed/bundle.py:87]: On a side note, I restored `dict` support in `Subcommand`'s argument handling (in https://github.com/bokeh/bokeh/pull/11290/commits/3f9e82d83e9a3590bedcb477830f74fb1a4c5488).

Comment (philippjfr) [reply bokeh/embed/bundle.py:87]: Definitely obscure but got copied over when I adapted some of the bundling code. Easy fix though.

Comment (philippjfr) [reply bokeh/embed/bundle.py:87]: And would never have caused real issues in a proper release since Panel is now strict in its pinning (i.e. it in pins `<2.4`)
